### DISCUSSION TITLE
Potential cause of slow dataloader on nemo main (1.11.0)

### DISCUSSION THREAD

**Discussion by feddybear** [NVIDIA-NeMo/NeMo#4524] [Q&A] [answered]
I am running the exact same Conformer-CTC training to compare in parallel, one using nemo 1.9 (PT lightning 1.5.10) and another using latest nemo version on main 1.11 (PT lightning 1.6.4).  The only difference between the default conf for Conformer-CTC is the addition of `model.skip_nan_grad` and `trainer.benchmark`, both set to `false`.  For 1.11, I tried training with and without these two additions.

The one running 1.9 is fine.

For the main (1.11) however, I'm stuck in the dataloader part and doesn't even show anything.  The last log is on the
```
Created a temporary directory at /tmp/<dir>
Writing /tmp/<dir>/_remote_module_non_sriptable.py
```
and based on the log for nemo 1.9, next is the loading of the training data and it's taking forever.  I can see some activity on `top` but it's almost an hour and nothing is happening.

What could be causing this, considering that I didn't change the original configuration of the default Conformer-CTC yaml file?

Any insight is greatly appreciated.  Thanks.

**Comment by feddybear** [comment, marked as answer]
nevermind, my bad, there's a subtle error on the container that still successfully built.

**Reply by titu1994** [reply]
For reference, what was the error ?

**Reply by feddybear** [reply]
@titu1994 , I didn't investigate further but it's related to the installation of `k2`, I think.  I will get a lot of killed "nvcc" lines on stderr.  The RAM requirements must be huge in the default configurations for locally building a docker image.  I didn't tinker with the number of jobs being set as I simply resorted to the images on ngc.

**Reply by titu1994** [reply]
@GNroy this might be problematic. Please look into it.

**Reply by GNroy** [reply]
k2 installation uses all available available threads to maximize build speed (see arg `-j` [here](https://github.com/NVIDIA/NeMo/blob/1428763b0045dc4a0d4783bcd8624a8bd46424fe/scripts/speech_recognition/k2/setup.sh#L23)). Last time I checked, k2 required <= 3gb per thread.
It is a bit surprising that triggering OOM while building an optional module can actually break the entire container.
Right now I don't have a good solution to this issue besides removing the k2 installation from the Dockerfile.

**Reply by titu1994** [reply]
Can't we limit threads ? 3 GB per thread, most local machines have 8 to 32 threads (physical or virtual), and surely most people do not have 64 or 128 GB ram. 

We might need to comment it out unless we have a way to not break container development.

**Reply by GNroy** [reply]
I thought about it at first. But there are two questions: "limit to what?" (we'd need to do some complex calculations of machine resources) and "what if the memory requirement increases again?"
I never thought I'd say this, but let's comment on it for now. Then, when time permits, we will make a binary installation possible (perhaps via a dedicated build server, a custom conda package, etc.).

**Reply by titu1994** [reply]
Can bash detect the num cpus and set it to possibly 4 or 8 ? If memory requirement increases, it will just slow down the compile right ? Or will it just hard crash ?

**Reply by titu1994** [reply]
If that won't work, let's comment it out. @fayejf when building the public container, we'll need to uncomment it and then build the container.
